include/kframework/lemmas/lemmas.k

```diff
@@ -76,16 +96,47 @@ module LEMMAS [symbolic]
     rule 0 <=Int keccak( _ )             => true [simplification]
     rule         keccak( _ ) <Int pow256 => true [simplification]
 
+    // Further keccak assumptions: these assumptions are not sound in principle, but are
+    // required for verification - they should be collected at the end of execution
+
+    // keccak does not equal a concrete value
+    rule [keccak-eq-conc-false]: keccak(_A)  ==Int _B => false [symbolic(_A), concrete(_B), simplification]
+    rule [keccak-neq-conc-true]: keccak(_A) =/=Int _B => true  [symbolic(_A), concrete(_B), simplification]
```

Comment:
Hmmmmm, these could be problematic. Say you have `keccak(X) ==Int SOME_HASH` in your constraints, but then somewhere else you have `X ==Int 3`, and then we have that `keccak(3) ==Int SOME_HASH` is true. This could happen, for example, if we looked up a symbolic storage location `X`, which later resolved to `3`, and `SOME_HASH` was the actual slot for storage location `3`.

But if these rules fire before the substitution `X ==Int 3` is applied and then `keccak` is simplified, we'll get the incorrect result. So it seems we're relying on the order that the backend tries things here.

Reply:
If they're useful for the engagement, then we should keep them there, but I don't think we should upstream them.